### PR TITLE
Fixing issues in Thread/Fiber merge

### PR DESCRIPTION
**Pull request by asksurya**
As discussed in https://github.com/ruby-prof/ruby-prof/issues/320#issuecomment-1469298989, the new solution has been implemented in these changes

### CODE REVIEW

File: test/call_tree_test.rb [hunk 1]
```diff
@@ -91,7 +91,7 @@ def test_add_child_gc
       GC.stress = false
     end
   end
-
+=begin
```
Comment (cfis) [inline test/call_tree_test.rb:94]: What is this for?

File: ext/ruby_prof/rp_call_tree.c [hunk 2]
```diff
@@ -2,9 +2,10 @@
    Please see the LICENSE file for copyright and distribution information */
 
 #include "rp_call_tree.h"
+#include "rp_call_trees.h"
```
Comment (cfis) [inline ext/ruby_prof/rp_call_tree.c:5]: This creates a circular reference between prof_call_tree_t and prof_call_trees_t. Can we avoid this?

File: ext/ruby_prof/rp_thread.c [hunk 3]
```diff
@@ -20,7 +20,7 @@ You cannot create an instance of RubyProf::Thread, instead you access it from a
 #include "rp_profile.h"
 
 VALUE cRpThread;
-
+st_table* merge_method_table;
```
Comment (cfis) [inline ext/ruby_prof/rp_thread.c:23]: This looks like global state to me, won't this conflict with multiple threads? 
If you need access to it, add a method to rp_thread.h:

```
st_table* thread_get_method_table(thread_data_t* thread);

File: ext/ruby_prof/rp_thread.c [hunk 4]
```diff
@@ -362,8 +362,12 @@ static VALUE prof_thread_merge(VALUE self, VALUE other)
 {
   thread_data_t* self_ptr = prof_get_thread(self);
   thread_data_t* other_ptr = prof_get_thread(other);
+  prof_method_table_merge(self_ptr->method_table,other_ptr->method_table);
+  merge_method_table= self_ptr->method_table;
```
Comment (cfis) [inline ext/ruby_prof/rp_thread.c:366]: Remove this. See comment above.

File: ext/ruby_prof/rp_thread.h [hunk 5]
```diff
@@ -6,7 +6,7 @@
 
 #include "ruby_prof.h"
 #include "rp_stack.h"
-
+extern st_table* merge_method_table;
```
Comment (cfis) [inline ext/ruby_prof/rp_thread.h:9]: Remove this.

**Comment by cfis** [comment]
Overall this looks pretty good - nice work! I added some comments for you to review. Also sync with latest master. You'll see I moved the merge tests to a new test file, including your original example. That test will need to be updated.

**Comment by asksurya** [comment]
sorry for the late response, there are some pointer issues causing seg faults in the code, i am debugging them. Will update this thread once i resolve them

**Comment by cfis** [comment]
No worries - thanks for working on this. You probably know this, but one thing I find helpful is:

```GC.stress = true```

You can put the into a setup method of a test or in the abstract test for all tests. It makes exceptions show up much more quickly instead of a few tests later which is really hard to debug.

**Comment by cfis** [comment]
Hi @asksurya just checking on this? Would be good to great to get this merged in!

**Comment by asksurya** [comment]
Hi @cfis ,
I have fixed the segmentation faults for this. I’ll implement the code changes and provide you the update by next week.

thanks,
Ashwin

File: ext/ruby_prof/rp_thread.c (hunk 3 above)
Comment (asivasub14) [reply ext/ruby_prof/rp_thread.c]: I have modified the methods now to pass the information from thread side into the call tree side and store it as a struct in the call_tree.c. I cannot do the thread.h becasue the call_tree is not having thread info and the struct is needed to call the merge_children code in rb_st_foreach

File: ext/ruby_prof/rp_thread.h (hunk 5 above)
Comment (asivasub14) [reply ext/ruby_prof/rp_thread.h]: Removed.

**Comment by asivasub14** [comment]
Hi @cfis,

I have updated the code logic to have the data correctly merged, it's the same logic suggested by you, I made some mistakes the first time around which are corrected now. W.R.T to the test code i will update them and submit again

Thanks,
Ashwin

**Comment by asksurya** [comment]
Hi @cfis,

I have fixed the tests as discussed as well as the changes suggested, Please let me know if anything is required further.

**Comment by cfis** [comment]
Great, I will take a look.

**Comment by cfis** [comment]
I rebased your branch on master, squished the commits together, then pushed to askuray branch - see https://github.com/ruby-prof/ruby-prof/commits/asksurya. Work will continue on that branch.

**Pull request merged**